File: data_structures/linked_list/linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    """
    Inserting operations
    """

    # Inserting a node at the start of the list
    def insertAtStart(self, data):
        new_node = Node(data=data)

        # if list is empty
        if self.head == None:
            self.head = new_node
            self.tail = new_node

        # if list is not empty
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

        self.length += 1


    # Inserting at the end of the list
    def InsertAtEnd(self, data):
        new_node = Node(data=data)

        # if list is empty
        if self.head == None:
            self.head = new_node
            self.tail = new_node
        
        # if list is not empty
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

    # Inserting at specific position
    def InsertAtPosition(self, data, position):

        pass
        


    def traverse(self):
        current = self.head
        while current is not None:
            print(current.data)
            current = current.next


l = LinkedList()
l.insertAtStart(1)
l.insertAtStart(2)
l.InsertAtEnd(4)
l.InsertAtPosition(10,2)
l.traverse()

File: java/linked_list/LinkedList.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class LinkedList {
    // uninitilized head node for refrence, this creates a refrence that can point to a node object but does not create an actual node
    Node head;

    // insert at start
    public void insertAtBeginning(int data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
    }

    // insert at end
    public void insertAtEnd(int data) {
        Node newNode = new Node(data);
        // check if list is empty
        if (head == null) {
            head = newNode;
        } else {
            // if list is not empty, then traverse the list and set last_node.next -> new_node
            Node temp = head;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newNode;
        }
    }

    // traverse the list
    public void traverse(){
        // check if list is empty
        if (head == null) {
            System.err.println("List is empty!");
        } else {
            Node temp = head;
            while (temp.next != null) {
                System.out.print(temp.data + "\n");
                temp = temp.next;
            }
        }
    }

    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        list.traverse();
        list.insertAtBeginning(0);
        list.insertAtBeginning(1);
        list.insertAtBeginning(2);
        list.insertAtBeginning(3);
        list.insertAtEnd(70);
        list.insertAtEnd(80);
        list.insertAtEnd(90);
        list.traverse();
    }

}
